### PR TITLE
[rfw] Add SliverPersistentHeader and SliverPersistentHeaderDelegate to RFW core widgets.

### PR DESCRIPTION
**Pull request by shuoch-g**
Add SliverPersistentHeader and SliverPersistentHeaderDelegate to RFW.

This is one example where RFW need to define its own logic when working with flutter widget (e,g SliverPersistentHeader in this PR). 

IMO, it is not always possible to remotely configure all properties of all widgets, `SliverPersistentHeaderDelegate` is one such example. This PR wants to demonstrate that RFW needs to come up with custom logic that is remotely configurable.

issue  #90218

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated `CHANGELOG.md` to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the `#hackers-new` channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by shuoch-g** [comment]
some discussion points about this PR:

1. I intentionally left out the "extension mechanism" for `SliverPersistentHeaderDelegate`, actually there can be more implementations of `SliverPersistentHeaderDelegate` than `FixedSliverPersistentHeaderDelegate` in this PR that can shrink the `child` and the shrink ratio can also be configured. I want to know owner's opinions on this before making such change.
2. I intentionally left out the tests, since `SliverPersistentHeader` need to be inside a `CustomScrollView`, which is also not there yet in RFW. I'd like to know owner's opinions on cases like `SliverPersistentHeaderDelegate` before adding `CustomScrollView`.

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 1]
```diff
@@ -779,6 +779,33 @@ class ArgumentDecoders {
...
+    switch (type) {
+      case 'fixed':
+        return FixedSliverPersistentHeaderDelegate(
+          fixedMaxExtent: source.v<double>([...key, 'maxExtent']) ?? 32.0, 
+          fixedMinExtent: source.v<double>([...key, 'minExtent']) ?? 0.0, 
+          child: source.child(['child']),
+        );
+      case null:
+      default:
+        return null;
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:805]: can we do something like the other extension mechanisms in this file?

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 2]
```diff
@@ -779,6 +779,33 @@ class ArgumentDecoders {
...
+  /// Returns a [SliverPersistentHeaderDelegate] from the specified map.
+  /// 
+  /// The `type` key specifies the kind of persistent header delegate.
+  /// 
+  /// Right now there is only one type `fixed` that creates a
+  /// [FixedSliverPersistentHeaderDelegate] using the keys `maxExtent`,
+  /// `minExtent` and `child`.
+  /// 
+  /// The types and defaults for these keys match the
+  /// identically named arguments to the default constructors of those classes.
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:791]:
```suggestion
  /// Right now there is only one type, `fixed`, that creates a
  /// [FixedSliverPersistentHeaderDelegate] using the keys `maxExtent`,
  /// `minExtent` and `child`.
  /// 
  /// The types and defaults for these keys match the
  /// identically-named arguments to the default constructors of those classes.
```

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 3]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
     }
   }
 }
+
+// TODO(shuoch-g): Move this class to other places where more appropriate.
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1451]: let's just create a new file to put them in

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 4]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+/// An implementation of [SliverPersistentHeaderDelegate] that takes in fixed
+/// child widget.
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1453]:
```suggestion
/// An implementation of [SliverPersistentHeaderDelegate] that takes in a fixed
/// child widget.
```

Comment (Hixie) [reply packages/rfw/lib/src/flutter/argument_decoders.dart:1453]: though i'm not really sure what this means exactly

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 5]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+/// 
+/// Caller can configure [maxExtent] and [minExtent], but the returned widget
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1455]: The caller...

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 6]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+/// in [build] is always [child] which is fixed (doesn't shrink).
+class FixedSliverPersistentHeaderDelegate
+    extends SliverPersistentHeaderDelegate {
+  /// Creates a [FixedSliverPersistentHeaderDelegate] delegate.
+  /// 
+  /// The [fixedMaxExtent], [fixedMinExtent] and [child] are all required,
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1461]: sentence is incomplete

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 7]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+  const FixedSliverPersistentHeaderDelegate({
+    required this.fixedMaxExtent,
+    required this.fixedMinExtent,
+    required this.child,
+  });
+
+  /// Fixed `maxExtent`.
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1468]: we should define what "fixed" means here and other places that use the term

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 8]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+  final double fixedMaxExtent;
+
+  /// Fixed `maxExtent`.
+  /// 
+  /// In [build], the [child] is fixed (not shrink according to this value).
+  final double fixedMinExtent;
+
+  /// Fixed widget to return in [build].
+  ///
+  /// In [build], it is fixed (not shrink).
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1480]: Not sure what "In [build], it is fixed (not shrink)." means

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 9]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+
+  @override
+  bool shouldRebuild(covariant SliverPersistentHeaderDelegate oldDelegate) =>
+    fixedMaxExtent != oldDelegate.maxExtent || 
+      fixedMinExtent != oldDelegate.minExtent;
+
+  // TODO(shuoch-g): Add fixed configurations below.
+  // FloatingHeaderSnapConfiguration
+  // OverScrollHeaderStretchConfiguration
+  // PersistentHeaderShowOnScreenConfiguration
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/argument_decoders.dart:1502]: we should either add these or just remove the comment here. There's an infinite number of things we don't yet support, no need to list them all. :-)

File: packages/rfw/lib/src/flutter/core_widgets.dart [hunk 10]
```diff
@@ -193,6 +194,9 @@ import 'runtime.dart';
 ///    to a [Rect] given by the `rect` key and a [TextDirection] given by the
 ///    `textDirection` key. New shader types can be added using
 ///    [ArgumentDecoders.shaderDecoders].
+/// 
+///  * the [SliverPersistentHeaderDelegate] type is represented as a map with a
+///    `type` key that can be `fixed` for [FixedSliverPersistentHeaderDelegate].
```
Comment (Hixie) [inline packages/rfw/lib/src/flutter/core_widgets.dart:199]: more detail here would be nice

Comment (Hixie) [reply packages/rfw/lib/src/flutter/core_widgets.dart:199]: (see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#writing-prompts-for-good-documentation)

**Comment by Hixie** [comment]
> some discussion points about this PR:
> 
> 1. I intentionally left out the "extension mechanism" for `SliverPersistentHeaderDelegate`, actually there can be more implementations of `SliverPersistentHeaderDelegate` than `FixedSliverPersistentHeaderDelegate` in this PR that can shrink the `child` and the shrink ratio can also be configured. I want to know owner's opinions on this before making such change.

I'm happy for us to add more here.


> 2. I intentionally left out the tests, since `SliverPersistentHeader` need to be inside a `CustomScrollView`, which is also not there yet in RFW. I'd like to know owner's opinions on cases like `SliverPersistentHeaderDelegate` before adding `CustomScrollView`.

Adding SliverPersistentHeaderDelegate seems perfectly reasonable.

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 1 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: Done. Thanks ;)

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 2 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: Done. Thanks ;)

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 4 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: Updated the doc comment, does it help now ?

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 5 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: Done.

File: packages/rfw/lib/src/flutter/argument_decoders.dart [hunk 11]
```diff
@@ -1420,3 +1447,57 @@ class ArgumentDecoders {
...
+    required this.fixedMinExtent,
+    required this.child,
+  });
+
+  /// Fixed `maxExtent`.
+  /// 
+  /// In [build], the [child] is fixed (not shrink according to this value).
+  final double fixedMaxExtent;
+
+  /// Fixed `maxExtent`.
```
[the remark this answers is not in the document]
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: Ooops, updated, thanks !

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 8 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: updated the doc, is it clear now ?

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 7 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: I just realized the `maxExtent` and `minExtent` should not prefix with `fixed`. Updated ;)

File: packages/rfw/lib/src/flutter/core_widgets.dart [hunk 12]
```diff
@@ -193,6 +194,9 @@ import 'runtime.dart';
 ///    to a [Rect] given by the `rect` key and a [TextDirection] given by the
 ///    `textDirection` key. New shader types can be added using
 ///    [ArgumentDecoders.shaderDecoders].
+/// 
+///  * the [SliverPersistentHeaderDelegate] type is represented as a map with a
```
[the remark this answers is not in the document]
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/core_widgets.dart]: Done.

**Comment by shuoch-g** [comment]
About 2, I prefer to add CustomScrollView in a separate PR, and making this PR as draft before that one is sent. Thanks !

File: packages/rfw/lib/src/flutter/argument_decoders.dart (hunk 9 above)
Comment (shuoch-g) [reply packages/rfw/lib/src/flutter/argument_decoders.dart]: removed them.

Yeah you are right. That's also why I would suggest we eventually keep this set of things in one place (RFW), it also benefits our project: it is much better if this set is in align with flutter team.

**Comment by stuartmorgan** [comment]
This has been in draft state for several months; is it something that is still being worked on?

**Comment by shuoch-g** [comment]
sorry about this, let me pick it up this week 😂

**Pull request closed without merging**